### PR TITLE
Adds search to supply console

### PR DESCRIPTION
**Pull request by Tastyfish**
## About The Pull Request

Every other civilized console has a search feature at this point. Now the supply console does too.

### Normal categories still work normally of course.
image

### Note that searching for fish also pull up Aquarium Kit
image

### Where's ma floor tiles??
image

The search checks both the name of the crate, and its description, and shows the first page's worth alphabetically, so we don't get weird scrolling nonsense.

Clicking away from the search clears it, and erasing the search term all the way, or clicking the little X when typing, bounces back to the first real category so you don't end up with shitty half-states.

I suppose this means that squeezing keywords into crate descriptions would be useful now.

## Why It's Good For The Game

No more customers squinting at the request console for 5 minutes trying to find beer.

## Changelog

:cl:
qol: The supply and supply request consoles now have a search bar.
/:cl:

### CODE REVIEW

**Comment by cacogen** [comment]
would be cool if it searched the contents of crates as well

**Comment by Tastyfish** [comment]
> would be cool if it searched the contents of crates as well

That would be neat, but I guess the main issue with that is it would create server-side overhead (to cache and then push the full list of items names) to benefit crates that have poor descriptions, as they should already be clear about what they contain (in theory).

**Comment by dragomagol** [comment]
Absolutely blessed PR

File: tgui/packages/tgui/interfaces/Cargo.js [hunk 1]
```diff
@@ -138,21 +140,52 @@ const CargoStatus = (props, context) => {
   );
 };
 
+/**
+ * Take entire supplies tree
+ * and return a flat supply pack list that matches search,
+ * sorted by name and only the first page.
+ * @param {category[]} supplies Supplies list.
+ * @param {string} search The search term
+ * @returns {pack[]} The flat list of supply packs.
```
Comment (stylemistake) [inline tgui/packages/tgui/interfaces/Cargo.js:149]: `category` and `pack` are not valid typescript types (they should be). So I guess, replace them with `any[]`.

**Review by stylemistake — APPROVED** [review]

File: tgui/packages/tgui/interfaces/Cargo.js (hunk 1 above)
Comment (Tastyfish) [reply tgui/packages/tgui/interfaces/Cargo.js:149]: Yeah, I was pretending this file was typescript (it really should be but that's for another time). I can make them `any[]` real quick.

Comment (Tastyfish) [reply tgui/packages/tgui/interfaces/Cargo.js]: Sorry, fixing them triggered a new review. The comment annotations are the only changes.

**Pull request merged**